### PR TITLE
[SPARK-36004][INFRA] Update MiMa and audit API changes

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to update MiMa based on Apache Spark 3.1.1 (the first release on 3.1 line) for Apache Spark 3.2.0 release.

### Why are the changes needed?

Old MiMa rules hides the breaking changes in Apache Spark 3.2.0. We need to audit and document it correctly in MiMa exclusion file. This issue is discussed here originally.
- https://github.com/apache/spark/pull/33196#issuecomment-873249068

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass the CIs

### CODE REVIEW

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -34,12 +34,42 @@ import com.typesafe.tools.mima.core.ProblemFilters._
...
+    ProblemFilters.exclude[Problem]("org.apache.spark.unused.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.unsafe.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.memory.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.util.collection.unsafe.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.sql.catalyst.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.sql.execution.*"),
+    ProblemFilters.exclude[Problem]("org.apache.spark.sql.internal.*"),
+
+    // [SPARK-34848][CORE] Add duration to TaskMetricDistributions
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.status.api.v1.TaskMetricDistributions.this"),
```
Comment (dongjoon-hyun) [inline project/MimaExcludes.scala:58]: cc @AngersZhuuuu and @srowen for SPARK-34848

File: project/MimaExcludes.scala [hunk 2]
```diff
@@ -34,12 +34,42 @@ import com.typesafe.tools.mima.core.ProblemFilters._
...
+
+    // [SPARK-34488][CORE] Support task Metrics Distributions and executor Metrics Distributions
+    // in the REST API call for a specified stage
+    ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.status.api.v1.StageData.this"),
```
Comment (dongjoon-hyun) [inline project/MimaExcludes.scala:62]: cc @AngersZhuuuu and @srowen for SPARK-34488

File: project/MimaExcludes.scala [hunk 3]
```diff
@@ -34,12 +34,42 @@ import com.typesafe.tools.mima.core.ProblemFilters._
...
+
+    // [SPARK-35896] Include more granular metrics for stateful operators in StreamingQueryProgress
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.sql.streaming.StateOperatorProgress.this"),
```
Comment (dongjoon-hyun) [inline project/MimaExcludes.scala:65]: cc @vkorukanti and @HeartSaVioR for SPARK-35896

**Comment by dongjoon-hyun** [comment]
Hi, All.
Although I checked the PR, I need your confirmation for this MiMa-detected change from 3.1.1.

cc @gengliangwang

File: project/MimaBuild.scala [hunk 4]
```diff
@@ -86,7 +86,7 @@ object MimaBuild {
 
   def mimaSettings(sparkHome: File, projectRef: ProjectRef): Seq[Setting[_]] = {
     val organization = "org.apache.spark"
-    val previousSparkVersion = "3.0.0"
+    val previousSparkVersion = "3.1.1"
```
Comment (dongjoon-hyun) [inline project/MimaBuild.scala:89]: This is intentionally the first version of 3.1 line. We don't have 3.1.0 and we don't need to change at every 3.1.x release.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by srowen** [comment]
I think the current state is that this passes only because some exclusions have to be retained that seem unintended - they were excluded for 3.0 - 3.1 not 3.1 - 3.2, but we were still using them. So I don't know if we should merge it ; could do but need to verify the exclusions are legitimate again or else undo the unintended api change

**Comment by srowen** [comment]
I think the exclusions are actually OK - all are private methods. If there's no other evidence to the contrary then this change is correct as-is.

**Review by LuciferYang — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen and all!

**Pull request closed without merging**